### ISSUE TITLE
kubeapi-load-balancer reports failed due to "ERROR juju.worker.dependency "metric-sender" manifold worker returned unexpected error: listen unix /var/lib/juju/agents/unit-kubeapi-load-balancer-0/425645480/s: socket: too many open files"

### ISSUE THREAD

**Issue by eric-sonchaiwanich** [juju-solutions/bundle-canonical-kubernetes#460]
Have experienced the kubeapi-load-balancer unit reporting a failed status due to the following error. 
 The load balancer itself continued to serve kubernetes api requests, however there were too many open files at the OS level.  

I seem to have found a workaround - see the end of this issue report.

`juju debug-log` showed the following:
```
unit-kubeapi-load-balancer-0: 15:37:45 ERROR juju.worker.dependency "metric-sender" manifold worker returned unexpected error: listen unix /var/lib/juju/agents/unit-kubeapi-load-balancer-0/425645480/s: socket: too many open files
unit-kubeapi-load-balancer-0: 15:37:46 DEBUG juju.worker.dependency "metric-collect" manifold worker stopped: failed to read charm from: /var/lib/juju/agents/unit-kubeapi-load-balancer-0/charm: open /var/lib/juju/agents/unit-kubeapi-load-balancer-0/charm/metadata.yaml: too many open files
unit-kubeapi-load-balancer-0: 15:37:46 ERROR juju.worker.dependency "metric-collect" manifold worker returned unexpected error: failed to read charm from: /var/lib/juju/agents/unit-kubeapi-load-balancer-0/charm: open /var/lib/juju/agents/unit-kubeapi-load-balancer-0/charm/metadata.yaml: too many open files
unit-kubeapi-load-balancer-0: 15:37:48 ERROR juju.juju.sockets failed to listen on unix:/var/lib/juju/agents/unit-kubeapi-load-balancer-0/585853671/s: listen unix /var/lib/juju/agents/unit-kubeapi-load-balancer-0/585853671/s: socket: too many open files
```

`sudo lsof | grep jujud`
``` 
jujud      1272 1498             root  988u     unix 0xffff88007a357800       0t0      39831 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/558102599/s type=STREAM
jujud      1272 1498             root  989u     unix 0xffff88007a357400       0t0      39838 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/294718202/s type=STREAM
jujud      1272 1498             root  990u     unix 0xffff88007a356400       0t0      39877 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/628507665/s type=STREAM
jujud      1272 1498             root  991u     unix 0xffff88007a356800       0t0      39886 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/809563196/s type=STREAM
jujud      1272 1498             root  992u     unix 0xffff88007a356000       0t0      39894 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/793540715/s type=STREAM
jujud      1272 1498             root  993u     unix 0xffff88007bbc0400       0t0      39912 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/259795662/s type=STREAM
jujud      1272 1498             root  994u     unix 0xffff88007bbc0000       0t0      39920 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/440497621/s type=STREAM
jujud      1272 1498             root  995u     unix 0xffff88007bbc0c00       0t0      39927 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/679595312/s type=STREAM
jujud      1272 1498             root  996u     unix 0xffff88007bbc1400       0t0      39937 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/209157839/s type=STREAM
jujud      1272 1498             root  997u     unix 0xffff88007bbc1800       0t0      39950 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/711819234/s type=STREAM
jujud      1272 1498             root  998u     unix 0xffff88007bbc1c00       0t0      39970 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/389866969/s type=STREAM
jujud      1272 1498             root  999u     unix 0xffff88007bbc1000       0t0      39994 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/750391652/s type=STREAM
jujud      1272 1498             root 1000u     unix 0xffff88007bbc2000       0t0      40003 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/303796595/s type=STREAM
jujud      1272 1498             root 1001u     unix 0xffff88007bbc0800       0t0      40018 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/685201974/s type=STREAM
jujud      1272 1498             root 1002u     unix 0xffff88007bbc2400       0t0      40024 /var/lib/juju/agents/unit-kubeapi-load-balancer-0/343290397/s type=STREAM
jujud
```

```
$ sudo lsof | grep jujud | wc -l
8440
```

Running the command every few seconds showed the number of open files growing (generally by 9 each few seconds):
```
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
733
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
733
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
742
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
751
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
769
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
778
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
778
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
787
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
787
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
787
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
796
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
796
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
805
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
805
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
805
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
814
ubuntu@ip-10-10-40-237:~$ sudo lsof | grep jujud | wc -l
814
```

/var/log/juju/unit-kubeapi-load-balancer-0.log showed
```
2017-12-21 15:46:47 DEBUG juju.worker.dependency engine.go:485 "metric-sender" manifold worker started
2017-12-21 15:46:47 DEBUG juju.worker.dependency engine.go:499 "metric-sender" manifold worker stopped: failed to open the metric reader: EOF
2017-12-21 15:46:47 ERROR juju.worker.dependency engine.go:546 "metric-sender" manifold worker returned unexpected error: failed to open the metric reader: EOF
2017-12-21 15:46:50 DEBUG juju.worker.dependency engine.go:485 "metric-sender" manifold worker started
2017-12-21 15:46:50 DEBUG juju.worker.dependency engine.go:499 "metric-sender" manifold worker stopped: failed to open the metric reader: EOF
2017-12-21 15:46:50 ERROR juju.worker.dependency engine.go:546 "metric-sender" manifold worker returned unexpected error: failed to open the metric reader: EOF
2017-12-21 15:46:53 DEBUG juju.worker.dependency engine.go:485 "metric-sender" manifold worker started
2017-12-21 15:46:53 DEBUG juju.worker.dependency engine.go:499 "metric-sender" manifold worker stopped: failed to open the metric reader: EOF
2017-12-21 15:46:53 ERROR juju.worker.dependency engine.go:546 "metric-sender" manifold worker returned unexpected error: failed to open the metric reader: EOF
2017-12-21 15:46:56 DEBUG juju.worker.dependency engine.go:485 "metric-sender" manifold worker started
2017-12-21 15:46:56 DEBUG juju.worker.dependency engine.go:499 "metric-sender" manifold worker stopped: failed to open the metric reader: EOF
2017-12-21 15:46:56 ERROR juju.worker.dependency engine.go:546 "metric-sender" manifold worker returned unexpected error: failed to open the metric reader: EOF
```

Details on environment:
```
Cloud/Region: aws/us-east-2
Version  2.2.2

App                    Version  Status  Scale  Charm                  Store       Rev  OS      Notes
canal                           active      6  canal                  jujucharms    4  ubuntu  
easyrsa                3.0.1    active      1  easyrsa                jujucharms   17  ubuntu  
etcd                   2.3.8    active      3  etcd                   jujucharms   50  ubuntu  
kubeapi-load-balancer  1.10.3   active    0/1  kubeapi-load-balancer  jujucharms   27  ubuntu  exposed
kubernetes-master      1.7.10   active      2  kubernetes-master      jujucharms   50  ubuntu  exposed
kubernetes-worker      1.7.10   active      4  kubernetes-worker      jujucharms   54  ubuntu  exposed
```

A resolution was found here:
https://ubuntuforums.org/showthread.php?t=2369476 - delete all files from `/var/lib/juju/metricspool/` directory.

**Comment by hyperbolic2346** [comment]
I saw this with an older juju as well. I think it was around 2.2.1. I upgraded to 2.3.1 and the problem went away. I assumed it was a fixed bug in juju.

**Comment by tvansteenburgh** [comment]
Yeah, this was fixed in juju 2.3: https://bugs.launchpad.net/juju/+bug/1733469
